### PR TITLE
Split release/dist_release makefile rules in two, reorganize and document makefiles

### PR DESCRIPTION
**Pull request by brson**
## What have you changed? (mandatory)

The main thing this patch does is to modify the default cargo release profile to use thinLTO instead of full LTO, and to turn debuginfo off. This reduces full build time by 25%, and partial build time by 30%.

It makes the various `release` and `dist_release` rules independent of each other. Now the `dist_release` rules rely on the `scripts/run-cargo.sh` script to put a special `.cargo/config` in place that contains the "real" release profile. This script requires the use of the unstable `-Zcargo-profile` flag.

One thing to note about this arrangement is that if any of the `dist_release` commands are interrupted they will leave `.cargo/config` in place. Subsequent cargo commands print a warning about cargo profiles being unstable, and effectively ignores the config file. I expect this to not be encountered by most contributors as they are not encouraged to run the `dist_release` targets by hand.

In the process it deletes all of the experimental `x-foo` makefile rules, leaving only one in place for `dist`.

As part of this work I restructured and documented the entire makefile. All pre-existing rules remain, they just might be in a different order. The `build_release` rule, which does the dist release without the extra testing and file moving, is renamed to `build_dist_release` for consistency, and `build_release` left in with a note to use `build_dist_release`.

I have confirmed @BusyJay that even with debuginfo off perf reports symbol names in profiles. Whether this is sufficient for practical profiling I don't know. We'll find out.

Despite weakening the LTO, I expect the runtime performance to still be excellent, and reasonable to use for benchmarking, as long as the comparison is done between two thinLTO builds. That means one should avoid comparing `release` builds to the actual released binaries; instead always make two `release` builds to compare, and only as a final sanity check, do `dist_release` builds.

Locally:

```
full before
    Finished release [optimized + debuginfo] target(s) in 29m 08s
partial before
    Finished release [optimized + debuginfo] target(s) in 18m 18s
full after
    Finished release [optimized] target(s) in 21m 51s
partial after
    Finished release [optimized] target(s) in 12m 41s
```

Fixes https://github.com/tikv/tikv/issues/4674

## What are the type of the changes? (mandatory)

- Improvement (change which is an improvement to an existing feature)

## How has this PR been tested? (mandatory)

Scripted timing tests and months of experiments.

## Does this PR affect documentation (docs) or release note? (mandatory)

No.

## Does this PR affect tidb-ansible update? (mandatory)

No.

## Refer to a related PR or issue link (optional)

https://github.com/tikv/tikv/issues/4674

## Benchmark result if necessary (optional)

See above.

### CODE REVIEW

**Comment by siddontang** [comment]
Great @brson 

Have we already fixed the problem that removing debuginfo from RocksDB?

File: Makefile [hunk 1]
```diff
@@ -54,86 +88,99 @@ default: release
...
+	@echo "the build_release rule no longer exists. you might want 'build_dist_release' or 'release'"
+	@exit 1
+
 # Distributable bins with SSE4.2 optimizations
 dist_unportable_release:
-	ROCKSDB_SYS_PORTABLE=0 make release
+	ROCKSDB_SYS_PORTABLE=0 make dist_release
 
 # Distributable bins with jemalloc memory profiling
 dist_prof_release:
```
Comment (siddontang) [inline Makefile]: I guess we can remove `dist_prof_release` and `dist_fail_release`, mostly we can't use them in production.

File: Makefile [hunk 2]
```diff
@@ -54,86 +88,99 @@ default: release
...
+# debuginfo. They take a very long time to build, so it is recommended
+# not to use them.
+
+# Build with release flag as if it were for distribution, but without
+# additional sanity checks and file movement.
+build_dist_release:
+	make x-build-dist
+
+# Temporary backwards compatibility
+build_release:
```
Comment (siddontang) [inline Makefile]: maybe we can remove this, it is rarely used.

Comment (brson) [reply Makefile]: Done.

File: Makefile (hunk 1 above)
Comment (brson) [reply Makefile]: Done.

**Comment by brson** [comment]
I pushed commits to address feedback @siddontang .

> Have we already fixed the problem that removing debuginfo from RocksDB?

No. There are two problems related to debuginfo in RocksDB. They aren't fixed, but they are also not related specifically to the release profile:

- https://github.com/tikv/tikv/issues/4711 - In dev mode, RocksDB always has debuginfo on. This is relatively easy to fix by hacking the rocksdb build script, and harder to fix properly in cmake-rs. I intend to fix this soon.

- https://github.com/tikv/tikv/issues/4894 - There isn't a single way to turn debuginfo back on for both Rust and C/C++. Right now this is addressed by educating developers to use RUSTFLAGS + CXXCFLAGS in the README. I don't know a good technical fix for this.

File: Makefile [hunk 3]
```diff
@@ -54,86 +88,85 @@ default: release
...
 dist_release:
-	make build_release
+	make build_dist_release
 	@mkdir -p ${BIN_PATH}
 	@cp -f ${CARGO_TARGET_DIR}/release/tikv-ctl ${CARGO_TARGET_DIR}/release/tikv-server ${CARGO_TARGET_DIR}/release/tikv-importer ${BIN_PATH}/
 	bash scripts/check-sse4_2.sh
 
+# Build with release flag as if it were for distribution, but without
+# additional sanity checks and file movement.
+build_dist_release:
```
Comment (siddontang) [inline Makefile]: @iamxy 

please make sure again that now we use dist_release to build TiKV in CI and publish this binary.

**Comment by siddontang** [comment]
I am with this change

PTAL @nrc @breeswish

File: Makefile [hunk 4]
```diff
@@ -153,6 +186,16 @@ test:
 	fi
 	bash scripts/check-bins-for-jemalloc.sh
 
+# Unlike test, this target will trace tests and output logs when fail
+# test is detected.  It's not clear who uses this and for what. It
+# does not seem to be used by CI. If you know please document it.
+trace_test:
```
Comment (Hoverbear) [inline Makefile]: Can we just remove it and re-add it if someone complains? I remember it was used in circleCI but haven't seen it used otherwise.

**Comment by Hoverbear** [comment]
This looks great, @brson , thanks for organizing this confusing file.

File: Makefile (hunk 4 above)
Comment (brson) [reply Makefile]: I have done so.

File: Makefile (hunk 3 above)
Comment (iamxy) [reply Makefile]: Yep, it has been confirmed.

**Pull request merged**